src/ICSharpCode.SharpZipLib/Checksum/Crc32Base.cs

```diff
@@ -0,0 +1,96 @@
+﻿using ICSharpCode.SharpZipLib.Checksum.Proxy;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ICSharpCode.SharpZipLib.Checksum
+{
+	/// <summary>
+	/// Base class for CRC32 checksums
+	/// </summary>
+	public abstract class Crc32Base : IChecksum
+	{
+		#region Instance Fields
+		/// <summary>
+		/// The CRC data checksum so far.
+		/// </summary>
+		private uint checkValue;
+
+		#endregion
+
+
+		/// <summary>
+		/// Initialise a default instance of <see cref="Crc32"></see>
+		/// </summary>
+		protected Crc32Base()
+		{
+			Reset();
+		}
+
+
+		internal abstract Crc32ProxyBase Proxy { get; }
+
+
+
+		/// <summary>
+		/// Resets the CRC data checksum as if no update was ever called.
+		/// </summary>
+		public void Reset()
+		{
+			checkValue = 0;
+		}
+
+		/// <summary>
+		/// Returns the CRC data checksum computed so far.
+		/// </summary>
+		/// <remarks>Reversed Out = false</remarks>
+		public long Value
+		{
+			get
+			{
+				return (long)(checkValue);
+			}
+		}
+
+		/// <summary>
+		/// Updates the checksum with the int bval.
+		/// </summary>
+		/// <param name = "bval">
+		/// the byte is taken as the lower 8 bits of bval
+		/// </param>
+		/// <remarks>Reversed Data = true</remarks>
+		public void Update(int bval)
+		{
+			checkValue = Proxy.Append(checkValue, new [] { (byte)bval }, 0, 1);
+		}
+
+
+
+		/// <summary>
+		/// Updates the CRC data checksum with the bytes taken from
+		/// a block of data.
+		/// </summary>
+		/// <param name="buffer">Contains the data to update the CRC with.</param>
+		public void Update(byte[] buffer)
+		{
+			if (buffer == null)
+			{
+				throw new ArgumentNullException(nameof(buffer));
+			}
+
+			Update(new ArraySegment<byte>(buffer, 0, buffer.Length));
```

Comment:
Pretty trivial, but just in case it's worth mentioning - is it worth creating an ArraySegment if the version of Update that takes an ArraySegment immediately unwraps it back into an array anyway?

Reply:
I did some benchmarks directly calling Proxy.Append, instead of calling the other version of Update. There is no speed benefit. But I do think the code is a bit inconsistent, there are 2 Update methods where Proxy.Append is called directly, then there is a 3rd Update method that calls one of the other 2 Update methods.

Reply:
I suppose that handling these sort of differences in where Span and friends can be useful, though I guess the extra dependency they have in versions of .Net prior to the latest would be unwanted?